protocols/kad/src/query/peers/disjoint_closest.rs

```diff
@@ -0,0 +1,427 @@
+    ///
+    /// If a successful result is not reported for a peer within this timeout
+    /// window, the iterator considers the peer unresponsive and will not wait for
+    /// the peer when evaluating the termination conditions, until and unless a
+    /// result is delivered. Defaults to `10` seconds.
+    pub peer_timeout: Duration,
+}
+
+impl Default for DisjointClosestPeersIterConfig {
+    fn default() -> Self {
+        DisjointClosestPeersIterConfig {
+            parallelism: ALPHA_VALUE.get(),
+            disjoint_paths: 3,
+            num_results: K_VALUE.get(),
+            peer_timeout: Duration::from_secs(10),
+        }
+    }
+}
+
+/// Wraps around a set of `ClosestPeersIter`, enforcing the amount of configured disjoint paths
+/// according to the S/Kademlia paper.
+pub struct DisjointClosestPeersIter {
+    iters: Vec<ClosestPeersIter>,
+    /// Mapping of yielded peers to iterator that yielded them.
+    ///
+    /// More specifically index into the `DisjointClosestPeersIter::iters` vector. On the one hand
+    /// this is used to link responses from remote peers back to the corresponding iterator, on the
+    /// other hand it is used to track which peers have been contacted in the past.
+    yielded_peers: HashMap<PeerId, usize>,
+}
+
+impl DisjointClosestPeersIter {
+    /// Creates a new iterator with a default configuration.
+    pub fn new<I>(target: KeyBytes, known_closest_peers: I) -> Self
+    where
+        I: IntoIterator<Item = Key<PeerId>>,
+    {
+        Self::with_config(
+            DisjointClosestPeersIterConfig::default(),
+            target,
+            known_closest_peers,
+        )
+    }
+
+    /// Creates a new iterator with the given configuration.
+    pub fn with_config<I, T>(
+        config: DisjointClosestPeersIterConfig,
+        target: T,
+        known_closest_peers: I,
+    ) -> Self
+    where
+        I: IntoIterator<Item = Key<PeerId>>,
+        T: Into<KeyBytes> + Clone,
+    {
+        let iters = split_inputs_per_disjoint_path(&config, known_closest_peers)
+            .into_iter()
+            .map(|(config, peers)| ClosestPeersIter::with_config(config, target.clone(), peers))
+            .collect();
+
+        DisjointClosestPeersIter {
+            iters,
+            yielded_peers: HashMap::new(),
+        }
+    }
+
+    pub fn on_failure(&mut self, peer: &PeerId) {
+        self.iters[self.yielded_peers[peer]].on_failure(peer);
+    }
+
+    pub fn on_success<I>(&mut self, peer: &PeerId, closer_peers: I)
+    where
+        I: IntoIterator<Item = PeerId>,
+    {
+        self.iters[self.yielded_peers[peer]].on_success(peer, closer_peers);
+    }
+
+    pub fn is_waiting(&self, peer: &PeerId) -> bool {
+        self.iters[self.yielded_peers[peer]].is_waiting(peer)
+    }
+
+    pub fn next(&mut self, now: Instant) -> PeersIterState {
+        let mut state = PeersIterState::Finished;
+
+        // TODO: Iterating through all iterators in the same order might be unfair.
```

Comment:
Is there no way to "select" on a bunch of iterators? That is, block until any iterator unblocks, and give me that one that just unblocked.

Reply:
None of this is asynchronous. You can think of the iterators of pure state machines with no side effects other than system time. All of their method calls are instant (ignoring CPU time).

Reply:
> In regards to any of the XPeersIterator: All of their methods are synchronous and single-threaded. While they are blocking in terms of not being asynchronous, 

OK so it is "logically" blocking from the perspective of the consumer of the iterator, that's what I meant.

However, the state machines need to accept both input from the consumer of the iterator, AND input from the network. I do not see anywhere in the code where the latter occurs.

For Kademlia specifically, we might receive an input from the network, e.g. the result of a previous hop, and then we want to react to this input, independently of whether the consumer of the iterator calls `next()` or not. Especially, we don't want to wait for another `next()` call before we respond to network inputs. But the way this code is written, is either hiding this part of the logic extremely well, or not performing it at all.

Reply:
> However, the state machines need to accept both input from the consumer of the iterator, AND input from the network. I do not see anywhere in the code where the latter occurs.

The iterator is not aware of what a _network_ is. It gets informed about occurred network events through its handlers [`on_success`](https://github.com/libp2p/rust-libp2p/pull/1473/files#diff-44f0d5820ec657b61472307702bcd0e7R98) and `on_failure`.

`on_success` can be called regardless of whether `next` was called recently, implying that handling of incoming events is happening outside the `next` calls.

Deciding when to reply to an input from the network and when to reach out to the next node is up to the `Kademlia` structure which owns a [`QueryPool`](https://github.com/libp2p/rust-libp2p/blob/master/protocols/kad/src/behaviour.rs#L59) which in turn owns e.g. a `DisjointClosestPeerIter`.

Reply:
Yes, I meant abstract input from an abstract networking. In general we have the following abstract inputs and outputs that (together with the algorithm state) should completely and deterministically affect the running of this algorithm:

- input from the network
- input from the local clock
- output to the network
- output to the consumer

This iterator implementation would seem to add another input:

- input from the consumer, requesting the "next" item in the iteration

This is nice though not a fundamental aspect of how the algorithm *must* work by definition.

Are you saying that `on_success` and `on_failure` represent the logic to handle input from the network (and presumably local clock, for timeouts)? That would seem to correspond to the `on_recv` function in my python code (granted, it does not deal with timeouts or failures). But the amount of code in these functions seems far too small to perform what (IMO) should be the logic.

Also, what is `is_waiting`?

It would help my understanding, if you could classify each of the existing functions as "input from X" or "output to X" like I did in the list above.

Reply:
> But the amount of code in these functions seems far too small to perform what (IMO) should be the logic.

Specifically, when one hop on one path succeeds, we want to choose the next hop based on (filtering out) stuff that was previously observed on other paths. But the current code doesn't do that. Most of the complex logic is in `next`, which is an input from the consumer of the iterator, not the abstract network.